### PR TITLE
[Neox 2.x] ImportBlocks add state root support

### PR DESCRIPTION
**Pull request by KickSeason**
* support state root export
* support state root import

@shargon @erikzhang

### CODE REVIEW

File: ImportBlocks/BlockImporter.cs [hunk 1]
```diff
@@ -49,6 +51,51 @@ private static IEnumerable<Block> GetBlocks(Stream stream, bool read_start = fal
             }
         }
 
+        private static IEnumerable<StateRoot> GetStateRoots(Stream stream)
+        {
+            using (BinaryReader r = new BinaryReader(stream))
+            {
+                uint start = r.ReadUInt32();
+                uint count = r.ReadUInt32();
+                uint end = start + count - 1;
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:60]: checked

File: ImportBlocks/BlockImporter.cs [hunk 2]
```diff
@@ -49,6 +51,51 @@ private static IEnumerable<Block> GetBlocks(Stream stream, bool read_start = fal
...
+                if (end <= Blockchain.Singleton.StateHeight) yield break;
+                for (uint height = start; height <= end; height++)
+                {
+                    byte[] array = r.ReadBytes(r.ReadInt32());
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:64]: Max length

File: ImportBlocks/BlockImporter.cs [hunk 3]
```diff
@@ -49,6 +51,51 @@ private static IEnumerable<Block> GetBlocks(Stream stream, bool read_start = fal
...
+            {
+                if (path.Start > Blockchain.Singleton.Height + 1) break;
+                if (path.IsCompressed)
+                    using (FileStream fs = new FileStream(path.FileName, FileMode.Open, FileAccess.Read, FileShare.Read))
+                    using (ZipArchive zip = new ZipArchive(fs, ZipArchiveMode.Read))
+                    using (Stream zs = zip.GetEntry(Path.GetFileNameWithoutExtension(path.FileName)).Open())
+                        foreach (var root in GetStateRoots(zs))
+                            yield return root;
+                else
+                    using (FileStream fs = new FileStream(path.FileName, FileMode.Open, FileAccess.Read, FileShare.Read))
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:93]: Share this using with `IsCompressed`

File: ImportBlocks/BlockImporter.cs (hunk 1 above)
Comment (KickSeason) [reply ImportBlocks/BlockImporter.cs:60]: check? line61?

**Comment by KickSeason** [comment]
> Maybe roots needs a plugin like statedumper that export continuously.

Good idea. I will build a plugin like that on neo3. We can use this first on neox-2.x.

**Comment by superboyiii** [comment]
It works well on mine.

File: ImportBlocks/BlockImporter.cs [hunk 4]
```diff
@@ -98,25 +147,49 @@ protected override void OnReceive(object message)
...
+                                rootsToImport.Add(stateRootsBeingImported.Current);
+                                continue;
+                            }
+                        }
+                        if (rootsToImport.Count > 0)
+                        {
+                            _blockchainActorRef.Tell(new Blockchain.ImportRoots { Roots = rootsToImport });
+                            break;
+                        }
+                        stateRootsBeingImported.Dispose();
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:190]:
```suggestion
                        stateRootsBeingImported.Dispose();
                        stateRootsBeingImported = null;
```

Please change the block import too

File: ImportBlocks/BlockImporter.cs [hunk 5]
```diff
@@ -98,25 +147,49 @@ protected override void OnReceive(object message)
...
                     blocksBeingImported = GetBlocksFromFile().GetEnumerator();
+                    stateRootsBeingImported = GetStateRootsFromFile().GetEnumerator();
                     // Start the first import
                     Self.Tell(new Blockchain.ImportCompleted());
                     break;
                 case Blockchain.ImportCompleted _:
                     // Import the next batch
-                    List<Block> blocksToImport = new List<Block>();
-                    for (int i = 0; i < BlocksPerBatch; i++)
+                    if (!(blocksBeingImported is null))
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:156]:
```suggestion
                    if (blocksBeingImported != null)
```
Please, change the block import too

**Comment by shargon** [comment]
I can edit, could you give me rights?

File: ImportBlocks/BlockImporter.cs [hunk 6]
```diff
@@ -98,25 +147,52 @@ protected override void OnReceive(object message)
...
+                        if (rootsToImport.Count > 0)
+                        {
+                            _blockchainActorRef.Tell(new Blockchain.ImportRoots { Roots = rootsToImport });
+                            break;
+                        }
+                        stateRootsBeingImported.Dispose();
                         blocksBeingImported.Dispose();
-                        _doneAction();
+                        stateRootsBeingImported = null;
+                        blocksBeingImported = null;
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:193]: This should be in line 172

**Pull request merged**